### PR TITLE
fix(preprocessing): apply correct h/w offsets to bounding boxes in RandomCrop

### LINKED ISSUE #23518: [Bug] RandomCrop shifts bounding box coordinates along the wrong axes (h/w swapped)
**System information**
- Keras version: 3.13.2
- Backend: tensorflow
- Environment: Google Colab (Python 3.13)

**Describe the bug**

`keras.layers.RandomCrop.transform_bounding_boxes` subtracts the **vertical** crop
offset `h_start` from the **x** coordinates and the **horizontal** offset `w_start`
from the **y** coordinates.

In `xyxy` format (`[x1, y1, x2, y2]`), x coordinates must be shifted by `w_start`
and y coordinates by `h_start`. The offsets are applied to the wrong axes, which
silently corrupts detection labels whenever `h_start != w_start` — i.e. virtually
every training step, since both offsets are drawn independently at random.

`CenterCrop` performs the equivalent shift correctly (x↔w, y↔h), which confirms
this is a bug and not a naming convention.

- Batched branch (rank-3 boxes):
  https://github.com/keras-team/keras/blob/master/keras/src/layers/preprocessing/image_preprocessing/random_crop.py#L237-L240
- Unbatched branch (rank-2 boxes):
  https://github.com/keras-team/keras/blob/master/keras/src/layers/preprocessing/image_preprocessing/random_crop.py#L247-L250

**Standalone code to reproduce the issue**

```python
import numpy as np
import keras

layer = keras.layers.RandomCrop(2, 4, bounding_box_format="xyxy")

out = layer.transform_bounding_boxes(
    {"boxes": np.array([[0.0, 0.0, 8.0, 8.0]]), "labels": np.array([0])},
    (3, 1),   # h_start=3, w_start=1: the crop starts at row 3, column 1
    training=True,
)
print(out["boxes"])
```
Expected behavior

The box is shifted horizontally by w_start=1 and vertically by h_start=3:

[[0. 0. 7. 5.]]   # x1=max(0-1,0), y1=max(0-3,0), x2=max(8-1,0), y2=max(8-3,0)
Actual behavior

tf.Tensor([[0. 0. 5. 7.]], shape=(1, 4), dtype=float32)   # x shifted by h_start, y shifted by w_start (swapped)
Impact

Object-detection models using RandomCrop with bounding boxes train on silently corrupted labels — no error, no warning. The existing test (test_dict_input in random_crop_test.py) only checks the output shape, never the coordinate values, so the bug ships unnoticed.

Related observations (possible follow-ups, not part of this report)

Only a lower bound (maximum(..., 0)) is applied; coordinates are not clipped to the crop size as CenterCrop does.
With relative formats (rel_xyxy, ...), the initial conversion to xyxy uses the crop dimensions (self.height/self.width) instead of the input image dimensions.
Would you like to help us fix it? Yes — happy to open a PR with the fix and value-based tests for both branches.

### PR DESCRIPTION
**Pull request by patminn107**
## Description

Fixes #23518

`RandomCrop.transform_bounding_boxes` subtracts the vertical crop offset
(`h_start`) from the x coordinates and the horizontal offset (`w_start`) from
the y coordinates. In `xyxy` format x coordinates must be shifted by `w_start`
and y coordinates by `h_start` — the offsets were applied to the wrong axes,
silently corrupting detection labels whenever `h_start != w_start` (i.e.
virtually every training step, since both offsets are drawn independently at
random).

`CenterCrop` performs the equivalent shift correctly (x↔w, y↔h).

## Changes

- `random_crop.py`: shift bounding boxes by the correct offsets and clip them
  to the crop size (`self.width` / `self.height`) so they stay within the
  cropped image, matching `CenterCrop` / `RandomZoom` behavior. Batched
  (rank-3) and unbatched (rank-2) boxes share a single code path via
  `boxes[..., i]` indexing.
- `random_crop_test.py`: add `test_bounding_boxes_offsets_applied_to_correct_axes`
  and `test_bounding_boxes_clipped_to_crop_size`, with explicit expected values
  for both branches. The existing bounding-box test only asserted shapes, which
  is why this shipped unnoticed.

## Before / after

Box `[3, 4, 5, 5]` (xyxy), crop offset `(h_start=3, w_start=1)`:

- Before this PR: `[0, 3, 2, 4]` (swapped axes)
- After this PR: `[2, 1, 4, 2]`

Box `[0, 0, 8, 8]` (xyxy), same crop offset:

- Before this PR: `[0, 0, 5, 7]` (swapped axes, extends beyond the 4x2 crop)
- After this PR: `[0, 0, 4, 2]` (correct axes, clipped to the crop size)

## Out of scope (follow-up)

- With relative formats (`rel_xyxy`, ...), the initial conversion to `xyxy`
  uses the crop dimensions instead of the input image dimensions.

## Tests

`keras/src/layers/preprocessing/image_preprocessing/random_crop_test.py`
passes with the tensorflow backend (Google Colab, CPU): 10 passed, including
the two new tests.

## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

### CODE REVIEW

File: keras/src/layers/preprocessing/image_preprocessing/random_crop.py [hunk 1]
```diff
@@ -234,20 +234,20 @@ def transform_bounding_boxes(
...
-                        self.backend.numpy.maximum(boxes[:, 1] - w_start, 0),
-                        self.backend.numpy.maximum(boxes[:, 2] - h_start, 0),
-                        self.backend.numpy.maximum(boxes[:, 3] - w_start, 0),
+                        self.backend.numpy.maximum(boxes[:, 0] - w_start, 0),
+                        self.backend.numpy.maximum(boxes[:, 1] - h_start, 0),
+                        self.backend.numpy.maximum(boxes[:, 2] - w_start, 0),
+                        self.backend.numpy.maximum(boxes[:, 3] - h_start, 0),
                     ],
                     axis=-1,
                 )
```
[the remark this answers is not in the document]
Comment (patminn107) [reply keras/src/layers/preprocessing/image_preprocessing/random_crop.py]: Thanks for the review! Applied both suggestions:
- Unified the batched/unbatched branches using `boxes[..., i]` indexing.
- Added upper-bound clipping to the crop size via `clip`, matching
  `CenterCrop` / `RandomZoom` behavior.

One deviation from the snippet: I used `self.backend.numpy.clip` /
`self.backend.numpy.stack` instead of the global `ops` module. This layer is
documented as safe inside `tf.data` pipelines, which relies on `self.backend`
being switched to the tensorflow backend inside a TF graph (`random_zoom.py`
uses the same pattern). Global `ops` would break `tf.data` usage when the
global backend is numpy/torch/jax.

Also added `test_bounding_boxes_clipped_to_crop_size`, covering both branches.

File: keras/src/layers/preprocessing/image_preprocessing/random_crop_test.py [hunk 2]
```diff
@@ -164,6 +164,37 @@ def test_dict_input(self):
...
+            "boxes": np.array([[[3.0, 4.0, 5.0, 5.0]]]),
+            "labels": np.array([[0]]),
+        }
+        output = layer.transform_bounding_boxes(
+            bounding_boxes, (3, 1), training=True
+        )
+        expected_boxes = np.array([[[2.0, 1.0, 4.0, 2.0]]])
+        self.assertAllClose(output["boxes"], expected_boxes)
+        self.assertAllClose(output["labels"], np.array([[0]]))
+
```
[the remark this answers is not in the document]
Comment (patminn107) [reply keras/src/layers/preprocessing/image_preprocessing/random_crop_test.py:197]: Done — `test_bounding_boxes_clipped_to_crop_size` covers a box extending
beyond the crop in both batched and unbatched forms.

**Review by hertschuh — CHANGES_REQUESTED** [review]
Thanks for the fix!

Can you run the code formatter on it to fix the code format error?

https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md#generating-public-api-and-formatting-the-code

**Comment by hertschuh** [comment]
The code format is still off somehow.

**Comment by patminn107** [comment]
Hi @hertschuh, I tried regenerating keras/api (and running the formatter) twice as you suggested, but both times the CI api-gen check still failed. Locally the regeneration looked deterministic and my local simulation passed, so I'm not sure what's causing the mismatch on CI (possibly a dependency-version drift between my environment and the CI image). I've reverted the regen commit since it wasn't helping.

How would you like to proceed ? should I leave keras/api as-is and let you regenerate it when merging, or is there something specific you'd like me to change? Thanks!

**Review by hertschuh — APPROVED** [review]
Thanks for the fix!
